core/WASM_core.md

```diff
@@ -0,0 +1,350 @@
+}
+
+struct HistogramRef {
+    host_ref: ExternRef,
+}
+
+struct GaugeRef {
+    host_ref: ExternRef,
+}
+
+impl CounterRef {
+    fn add(&self, value: u64, attributes: MetricsAttributesRef) {}
+}
+
+impl HistogramRef {
+    // When referring to durations, this value is in millis
+    fn record(&self, value: u64, attributes: MetricsAttributesRef) {}
+}
+
+impl GaugeRef {
+    fn record(&self, value: u64, attributes: MetricsAttributesRef) {}
+}
+
+struct Buckets {
+    buckets: Vec<f64>,
+}
+
+// -------------- Logging --------------
+
+// In Core, all logging is represented as events on trace spans. However, adding a full tracing
+// API would be insanely huge, and way, way too much work on both sides. Additionally, OTel simply
+// does not seem designed for exporting and importing traces in the same process (I can't find any
+// in-memory importers). Even if that worked, just defining the Span struct is significant. All
+// that in mind, I propose only adding a logging interface here.
+
+// This is copied straight out of Core today
+pub struct CoreLog {
+    /// The module within core this message originated from
+    pub target: String,
+    /// Log message
+    pub message: String,
+    /// Time log was generated
+    pub timestamp: SystemTime,
+    /// Message level
+    pub level: Level,
+    /// Arbitrary k/v pairs (span k/vs are collapsed with event k/vs here).
+    pub fields: HashMap<String, serde_json::Value>,
+    /// A list of the outermost to the innermost span names
+    pub span_contexts: Vec<String>,
+}
+
+enum Level {
+    Trace,
+    Debug,
+    Info,
+    Warn,
+    Error,
+}
+
+// The same assumption about a worker-per-module-instance applies here as with metrics.
+fn log(log: CoreLog) {}
+```
+
+#### An aside on serialization
+
+We are presented with an opportunity here to get rid of the ser/de overhead currently present
+as a result of using protos for everything. Since we are introducing additional overhead b/c of
+WASM, we could maybe introduce some savings here. My **completely unsubstantiated** guess is that if
+we stopped the unneeded ser/de we might actually get slightly *faster* under WASM which would be
+amusing.
+
+The options here are Cap'n'Proto or Flatbuffers. Both appear to have support for all the languages
+we care about, except for Flatbuffers not having a Ruby implementation.
+
+Flatbuffer's design is more oriented towards what we're doing, whereas Cap'n'Proto is more concerned
+with sending things over the wire, but is perhaps a bit more ergonomic.
+
+In either case, the primary pain point would be figuring out how to deal with the annoyance of
+having lifetime annotations in the Rust code. Cap'n'Proto has built-in support for creating owned
+versions of the structs which makes that pretty easy, but at that point it's unclear if whatever
+speedup we'd get is worth it compared to Protobuf.
+
+Ultimately, it may makes sense to defer this work until we have measured performance and decided
```

Comment:
It's already going to be painful to move away from protobuf having 4 Lang implementations, I'd love for us to do this optimization but I feel like it's not time yet and shouldn't be coupled to this effort.

Reply:
That doesn't really get at my more important second point, though, which is I think we should at least try this out and see what the perf difference is. If it's hugely substantial, we probably want to just pay this cost now, because changing later is more expensive than doing it together.

Reply:
I think this requires discussion because of the required level of investment.